File: demos/python/basic_proxy.py
#!/usr/bin/env python3
"""
Outbound LB - Python Demo

Demonstrates how to use Outbound LB proxy with Python using the requests library.
Includes examples for HTTP, HTTPS, authentication, error handling, and concurrent requests.

Requirements:
    pip install requests aiohttp

Usage:
    python basic_proxy.py
"""

import os
import sys
import asyncio
from concurrent.futures import ThreadPoolExecutor, as_completed

# Configuration from environment variables
PROXY_HOST = os.getenv("PROXY_HOST", "localhost")
PROXY_PORT = os.getenv("PROXY_PORT", "3128")
PROXY_USER = os.getenv("PROXY_USER", "user")
PROXY_PASS = os.getenv("PROXY_PASS", "password")

# Proxy URLs
PROXY_URL = f"http://{PROXY_HOST}:{PROXY_PORT}"
PROXY_URL_AUTH = f"http://{PROXY_USER}:{PROXY_PASS}@{PROXY_HOST}:{PROXY_PORT}"


def example_http_request():
    """Basic HTTP request through the proxy."""
    import requests

    print("=" * 60)
    print("Example 1: Basic HTTP Request")
    print("=" * 60)

    proxies = {
        "http": PROXY_URL,
        "https": PROXY_URL,
    }

    try:
        response = requests.get("http://httpbin.org/ip", proxies=proxies, timeout=10)
        print(f"Status: {response.status_code}")
        print(f"Response: {response.json()}")
    except requests.RequestException as e:
        print(f"Error: {e}")

    print()


def example_https_request():
    """HTTPS request through the proxy (CONNECT tunnel)."""
    import requests

    print("=" * 60)
    print("Example 2: HTTPS Request (CONNECT tunnel)")
    print("=" * 60)

    proxies = {
        "http": PROXY_URL,
        "https": PROXY_URL,
    }

    try:
        response = requests.get("https://httpbin.org/ip", proxies=proxies, timeout=10)
        print(f"Status: {response.status_code}")
        print(f"Response: {response.json()}")
    except requests.RequestException as e:
        print(f"Error: {e}")

    print()


def example_authenticated_proxy():
    """Request through authenticated proxy."""
    import requests

    print("=" * 60)
    print("Example 3: Authenticated Proxy")
    print("=" * 60)

    proxies = {
        "http": PROXY_URL_AUTH,
        "https": PROXY_URL_AUTH,
    }

    try:
        response = requests.get("https://httpbin.org/ip", proxies=proxies, timeout=10)
        print(f"Status: {response.status_code}")
        print(f"Response: {response.json()}")
    except requests.RequestException as e:
        print(f"Error: {e}")

    print()


def example_error_handling():
    """Demonstrate error handling for common scenarios."""
    import requests

    print("=" * 60)
    print("Example 4: Error Handling")
    print("=" * 60)

    proxies = {
        "http": PROXY_URL,
        "https": PROXY_URL,
    }

    # Test connection timeout
    print("Testing connection timeout...")
    try:
        response = requests.get(
            "http://httpbin.org/delay/5",
            proxies=proxies,
            timeout=2,  # Short timeout to trigger error
        )
    except requests.Timeout:
        print("  Caught timeout error (expected)")
    except requests.RequestException as e:
        print(f"  Error: {e}")

    # Test invalid URL
    print("Testing invalid URL...")
    try:
        response = requests.get(
            "http://invalid.invalid.invalid",
            proxies=proxies,
            timeout=5,
        )
    except requests.RequestException as e:
        print(f"  Caught connection error (expected): {type(e).__name__}")

    # Test proxy authentication failure (if auth is required)
    print("Testing proxy auth failure (if auth required on proxy)...")
    try:
        bad_proxy = f"http://wronguser:wrongpass@{PROXY_HOST}:{PROXY_PORT}"
        response = requests.get(
            "http://httpbin.org/ip",
            proxies={"http": bad_proxy, "https": bad_proxy},
            timeout=5,
        )
        if response.status_code == 407:
            print("  Got 407 Proxy Authentication Required (expected)")
        else:
            print(f"  Status: {response.status_code}")
    except requests.RequestException as e:
        print(f"  Error: {e}")

    print()


def example_concurrent_requests():
    """Demonstrate concurrent requests to show load balancing."""
    import requests

    print("=" * 60)
    print("Example 5: Concurrent Requests (Load Balancing Demo)")
    print("=" * 60)

    proxies = {
        "http": PROXY_URL,
        "https": PROXY_URL,
    }

    def make_request(request_id):
        """Make a single request and return the source IP."""
        try:
            response = requests.get(
                "http://httpbin.org/ip",
                proxies=proxies,
                timeout=10,
            )
            return request_id, response.json().get("origin", "unknown")
        except requests.RequestException as e:
            return request_id, f"error: {e}"

    num_requests = 10
    print(f"Making {num_requests} concurrent requests...")

    # Track IP distribution
    ip_counts = {}

    with ThreadPoolExecutor(max_workers=5) as executor:
        futures = [executor.submit(make_request, i) for i in range(num_requests)]

        for future in as_completed(futures):
            request_id, ip = future.result()
            print(f"  Request {request_id}: {ip}")

            if not ip.startswith("error"):
                ip_counts[ip] = ip_counts.get(ip, 0) + 1

    print("\nIP Distribution:")
    for ip, count in sorted(ip_counts.items()):
        print(f"  {ip}: {count} requests")

    print()


async def example_async_requests():
    """Demonstrate async requests using aiohttp."""
    try:
        import aiohttp
    except ImportError:
        print("=" * 60)
        print("Example 6: Async Requests (aiohttp)")
        print("=" * 60)
        print("Skipped: aiohttp not installed")
        print("Install with: pip install aiohttp")
        print()
        return

    print("=" * 60)
    print("Example 6: Async Requests (aiohttp)")
    print("=" * 60)

    async def fetch(session, url, request_id):
        """Make an async request."""
        try:
            async with session.get(url, timeout=aiohttp.ClientTimeout(total=10)) as response:
                data = await response.json()
                return request_id, data.get("origin", "unknown")
        except Exception as e:
            return request_id, f"error: {e}"

    num_requests = 10
    print(f"Making {num_requests} async requests...")

    connector = aiohttp.TCPConnector(limit=10)

    async with aiohttp.ClientSession(
        connector=connector,
        trust_env=False,  # Don't use system proxy settings
    ) as session:
        # Set proxy for aiohttp
        session._connector._proxy = PROXY_URL

        tasks = [
            fetch(session, "http://httpbin.org/ip", i) for i in range(num_requests)
        ]

        results = await asyncio.gather(*tasks)

        ip_counts = {}
        for request_id, ip in results:
            print(f"  Request {request_id}: {ip}")
            if not ip.startswith("error"):
                ip_counts[ip] = ip_counts.get(ip, 0) + 1

        print("\nIP Distribution:")
        for ip, count in sorted(ip_counts.items()):
            print(f"  {ip}: {count} requests")

    print()


def main():
    """Run all examples."""
    print()
    print("Outbound LB - Python Demo")
    print(f"Proxy: {PROXY_URL}")
    print()

    # Check if requests is available
    try:
        import requests
    except ImportError:
        print("Error: 'requests' library is required")
        print("Install with: pip install requests")
        sys.exit(1)

    # Run examples
    example_http_request()
    example_https_request()
    example_authenticated_proxy()
    example_error_handling()
    example_concurrent_requests()

    # Run async example
    asyncio.run(example_async_requests())

    print("All examples completed!")


if __name__ == "__main__":
    main()

File: demos/nodejs/basic_proxy.js
#!/usr/bin/env node
/**
 * Outbound LB - Node.js Demo
 *
 * Demonstrates how to use Outbound LB proxy with Node.js using axios and https-proxy-agent.
 * Includes examples for HTTP, HTTPS, authentication, error handling, and concurrent requests.
 *
 * Requirements:
 *     npm install axios https-proxy-agent http-proxy-agent
 *
 * Usage:
 *     node basic_proxy.js
 */

const http = require('http');
const https = require('https');

// Configuration from environment variables
const PROXY_HOST = process.env.PROXY_HOST || 'localhost';
const PROXY_PORT = process.env.PROXY_PORT || '3128';
const PROXY_USER = process.env.PROXY_USER || 'user';
const PROXY_PASS = process.env.PROXY_PASS || 'password';

const PROXY_URL = `http://${PROXY_HOST}:${PROXY_PORT}`;
const PROXY_URL_AUTH = `http://${PROXY_USER}:${PROXY_PASS}@${PROXY_HOST}:${PROXY_PORT}`;

// Try to load optional dependencies
let axios, HttpsProxyAgent, HttpProxyAgent;

try {
    axios = require('axios');
    HttpsProxyAgent = require('https-proxy-agent').HttpsProxyAgent;
    HttpProxyAgent = require('http-proxy-agent').HttpProxyAgent;
} catch (e) {
    console.error('Dependencies not installed. Run: npm install axios https-proxy-agent http-proxy-agent');
    process.exit(1);
}

/**
 * Example 1: Basic HTTP Request
 */
async function exampleHttpRequest() {
    console.log('='.repeat(60));
    console.log('Example 1: Basic HTTP Request');
    console.log('='.repeat(60));

    try {
        const agent = new HttpProxyAgent(PROXY_URL);
        const response = await axios.get('http://httpbin.org/ip', {
            httpAgent: agent,
            proxy: false, // Important: disable axios's default proxy handling
            timeout: 10000,
        });
        console.log(`Status: ${response.status}`);
        console.log(`Response: ${JSON.stringify(response.data)}`);
    } catch (error) {
        console.log(`Error: ${error.message}`);
    }

    console.log();
}

/**
 * Example 2: HTTPS Request (CONNECT tunnel)
 */
async function exampleHttpsRequest() {
    console.log('='.repeat(60));
    console.log('Example 2: HTTPS Request (CONNECT tunnel)');
    console.log('='.repeat(60));

    try {
        const agent = new HttpsProxyAgent(PROXY_URL);
        const response = await axios.get('https://httpbin.org/ip', {
            httpsAgent: agent,
            proxy: false,
            timeout: 10000,
        });
        console.log(`Status: ${response.status}`);
        console.log(`Response: ${JSON.stringify(response.data)}`);
    } catch (error) {
        console.log(`Error: ${error.message}`);
    }

    console.log();
}

/**
 * Example 3: Authenticated Proxy
 */
async function exampleAuthenticatedProxy() {
    console.log('='.repeat(60));
    console.log('Example 3: Authenticated Proxy');
    console.log('='.repeat(60));

    try {
        const agent = new HttpsProxyAgent(PROXY_URL_AUTH);
        const response = await axios.get('https://httpbin.org/ip', {
            httpsAgent: agent,
            proxy: false,
            timeout: 10000,
        });
        console.log(`Status: ${response.status}`);
        console.log(`Response: ${JSON.stringify(response.data)}`);
    } catch (error) {
        console.log(`Error: ${error.message}`);
    }

    console.log();
}

/**
 * Example 4: Error Handling
 */
async function exampleErrorHandling() {
    console.log('='.repeat(60));
    console.log('Example 4: Error Handling');
    console.log('='.repeat(60));

    // Test connection timeout
    console.log('Testing connection timeout...');
    try {
        const agent = new HttpsProxyAgent(PROXY_URL);
        await axios.get('http://httpbin.org/delay/5', {
            httpsAgent: agent,
            proxy: false,
            timeout: 2000, // Short timeout
        });
    } catch (error) {
        if (error.code === 'ECONNABORTED' || error.message.includes('timeout')) {
            console.log('  Caught timeout error (expected)');
        } else {
            console.log(`  Error: ${error.message}`);
        }
    }

    // Test invalid URL
    console.log('Testing invalid URL...');
    try {
        const agent = new HttpsProxyAgent(PROXY_URL);
        await axios.get('http://invalid.invalid.invalid', {
            httpsAgent: agent,
            proxy: false,
            timeout: 5000,
        });
    } catch (error) {
        console.log(`  Caught connection error (expected): ${error.code || error.message}`);
    }

    // Test proxy authentication failure
    console.log('Testing proxy auth failure (if auth required on proxy)...');
    try {
        const badProxyUrl = `http://wronguser:wrongpass@${PROXY_HOST}:${PROXY_PORT}`;
        const agent = new HttpsProxyAgent(badProxyUrl);
        const response = await axios.get('http://httpbin.org/ip', {
            httpsAgent: agent,
            httpAgent: new HttpProxyAgent(badProxyUrl),
            proxy: false,
            timeout: 5000,
            validateStatus: () => true, // Accept any status code
        });
        if (response.status === 407) {
            console.log('  Got 407 Proxy Authentication Required (expected)');
        } else {
            console.log(`  Status: ${response.status}`);
        }
    } catch (error) {
        console.log(`  Error: ${error.message}`);
    }

    console.log();
}

/**
 * Example 5: Concurrent Requests (Load Balancing Demo)
 */
async function exampleConcurrentRequests() {
    console.log('='.repeat(60));
    console.log('Example 5: Concurrent Requests (Load Balancing Demo)');
    console.log('='.repeat(60));

    const numRequests = 10;
    console.log(`Making ${numRequests} concurrent requests...`);

    const ipCounts = {};

    const makeRequest = async (requestId) => {
        try {
            const agent = new HttpProxyAgent(PROXY_URL);
            const response = await axios.get('http://httpbin.org/ip', {
                httpAgent: agent,
                proxy: false,
                timeout: 10000,
            });
            return { requestId, ip: response.data.origin };
        } catch (error) {
            return { requestId, ip: `error: ${error.message}` };
        }
    };

    const promises = [];
    for (let i = 0; i < numRequests; i++) {
        promises.push(makeRequest(i));
    }

    const results = await Promise.all(promises);

    for (const { requestId, ip } of results) {
        console.log(`  Request ${requestId}: ${ip}`);
        if (!ip.startsWith('error')) {
            ipCounts[ip] = (ipCounts[ip] || 0) + 1;
        }
    }

    console.log('\nIP Distribution:');
    for (const [ip, count] of Object.entries(ipCounts).sort()) {
        console.log(`  ${ip}: ${count} requests`);
    }

    console.log();
}

/**
 * Example 6: Using native HTTP module
 */
async function exampleNativeHttp() {
    console.log('='.repeat(60));
    console.log('Example 6: Using Native HTTP Module');
    console.log('='.repeat(60));

    return new Promise((resolve) => {
        const options = {
            hostname: PROXY_HOST,
            port: parseInt(PROXY_PORT),
            path: 'http://httpbin.org/ip',
            method: 'GET',
            headers: {
                'Host': 'httpbin.org',
            },
        };

        const req = http.request(options, (res) => {
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                console.log(`Status: ${res.statusCode}`);
                try {
                    console.log(`Response: ${data}`);
                } catch (e) {
                    console.log(`Response: ${data}`);
                }
                console.log();
                resolve();
            });
        });

        req.on('error', (error) => {
            console.log(`Error: ${error.message}`);
            console.log();
            resolve();
        });

        req.setTimeout(10000, () => {
            req.destroy();
            console.log('Request timeout');
            console.log();
            resolve();
        });

        req.end();
    });
}

/**
 * Main function
 */
async function main() {
    console.log();
    console.log('Outbound LB - Node.js Demo');
    console.log(`Proxy: ${PROXY_URL}`);
    console.log();

    await exampleHttpRequest();
    await exampleHttpsRequest();
    await exampleAuthenticatedProxy();
    await exampleErrorHandling();
    await exampleConcurrentRequests();
    await exampleNativeHttp();

    console.log('All examples completed!');
}

main().catch(console.error);
